### PR TITLE
[FLINK-6177] Add support for "Distributed Cache" in streaming applica…

### PR DESCRIPTION
**Pull request by zohar-mizrahi**
…tions

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by StephanEwen** [comment]
Thanks for contributing this, the added functionality looks good.

I would prefer to add this change without changing the dependencies and test base classes. You could for example change the test to throw an exception in the "validator function" if the word is not in the cache file. That way you do not need to "collect back" the data.

Minor comment: Generating the input from a collection rather than a file makes the tests usually a bit more lightweight. In all newer tests, we try to do that.

**Comment by zohar-pm** [comment]
No problem. Will follow it and will create a new pull request.

**Comment by zentol** [comment]
@zohar-pm You don't have to open a new one, feel free to update the branch in this one.

**Comment by StephanEwen** [comment]
Looks quite good now.

If I can ask you for one more followup: To have faster tests, it would be good to add the streaming distributed cache test and the batch distributed cache test to the same file.

Can you change the `DistributedCacheTest` to extend `StreamingMultipleProgramsTestBase` and put your test in there as well? That will cause only one distributed mini cluster to be spawned, which typically saves 1-2 secs in tests. May not seem much, but it adds up over the 1000s of tests Flink has by now...

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -1776,8 +1787,45 @@ public static void setDefaultLocalParallelism(int parallelism) {
...
+	 * <p>
+	 * The {@link org.apache.flink.api.common.functions.RuntimeContext} can be obtained inside UDFs via
+	 * {@link org.apache.flink.api.common.functions.RichFunction#getRuntimeContext()} and provides access
+	 * {@link org.apache.flink.api.common.cache.DistributedCache} via
+	 * {@link org.apache.flink.api.common.functions.RuntimeContext#getDistributedCache()}.
+	 *
+	 * @param filePath The path of the file, as a URI (e.g. "file:///some/path" or "hdfs://host:port/and/path")
+	 * @param name The name under which the file is registered.
+	 */
+	public void registerCachedFile(String filePath, String name){
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: missing space before `{`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -1776,8 +1787,45 @@ public static void setDefaultLocalParallelism(int parallelism) {
...
+	 * The {@link org.apache.flink.api.common.functions.RuntimeContext} can be obtained inside UDFs via
+	 * {@link org.apache.flink.api.common.functions.RichFunction#getRuntimeContext()} and provides access
+	 * {@link org.apache.flink.api.common.cache.DistributedCache} via
+	 * {@link org.apache.flink.api.common.functions.RuntimeContext#getDistributedCache()}.
+	 *
+	 * @param filePath The path of the file, as a URI (e.g. "file:///some/path" or "hdfs://host:port/and/path")
+	 * @param name The name under which the file is registered.
+	 * @param executable flag indicating whether the file should be executable
+	 */
+	public void registerCachedFile(String filePath, String name, boolean executable){
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: missing space before `}`.

**Comment by zentol** [comment]
merging. will add the missing space while I'm doing it.

**Pull request closed without merging**